pkg/controller/signer/signer-controller.go

```diff
@@ -0,0 +1,269 @@
+	// TODO: We may need a more robust way to do this later
+	csr.Status.Conditions = append(csr.Status.Conditions, csrv1.CertificateSigningRequestCondition{
+		Type:    csrv1.CertificateApproved,
+		Status:  "True",
+		Reason:  "AutoApproved",
+		Message: "Automatically approved by " + signerName})
+	// Update status to "Approved"
+	csr, err = r.clientset.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), request.Name, csr, metav1.UpdateOptions{})
+	if err != nil {
+		signerFailure(r, csr, "UpdateApprovalFailure",
+			fmt.Sprintf("Could not update CSR status to approved: %v", err))
+		return reconcile.Result{}, err
+	}
+
+	// Get our CA that was created by the operatorpki.
+	caSecret := &corev1.Secret{}
+	err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "openshift-ovn-kubernetes", Name: "signer-ca"}, caSecret)
+	if err != nil {
+		signerFailure(r, csr, "CAFailure",
+			fmt.Sprintf("Could not get CA certificate and key: %v", err))
+		return reconcile.Result{}, err
+	}
+
+	// Decode the certificate request from PEM format.
+	certReq, err := decodeCertificateRequest(csr.Spec.Request)
+	if err != nil {
+		// We dont degrade the status of the controller as this is due to a
+		// malformed CSR rather than an issue with the controller.
+		updateCSRStatusConditions(r, csr, "CSRDecodeFailure",
+			fmt.Sprintf("Could not decode Certificate Request: %v", err))
+		return reconcile.Result{}, nil
+	}
+
+	// Decode the CA certificate from PEM format.
+	caCert, err := decodeCertificate(caSecret.Data["tls.crt"])
+	if err != nil {
+		signerFailure(r, csr, "CorruptCACert",
+			fmt.Sprintf("Unable to decode CA certificate for %v: %v", signerName, err))
+		return reconcile.Result{}, nil
+	}
+
+	// Decode the CA key from PEM format.
+	caKey, err := decodePrivateKey(caSecret.Data["tls.key"])
+	if err != nil {
+		signerFailure(r, csr, "CorruptCAKey",
+			fmt.Sprintf("Unable to decode CA private key for %v: %v", signerName, err))
+		return reconcile.Result{}, nil
+	}
+
+	// Create a new certificate using the certificate template and certificate.
+	// We can then sign this using the CA.
+	signedCert, err := signCSR(newCertificateTemplate(certReq), certReq.PublicKey, caCert, caKey)
+	if err != nil {
+		signerFailure(r, csr, "SigningFailure",
+			fmt.Sprintf("Unable to sign certificate for %v and signer %v: %v", request.Name, signerName, err))
+		return reconcile.Result{}, nil
+	}
+
+	// Encode the certificate into PEM format and add to the status of the CSR
+	csr.Status.Certificate, err = crypto.EncodeCertificates(signedCert)
+	if err != nil {
+		signerFailure(r, csr, "EncodeFailure",
+			fmt.Sprintf("Could not encode certificate: %v", err))
+		return reconcile.Result{}, nil
+	}
+
+	err = r.client.Status().Update(context.TODO(), csr)
+	if err != nil {
+		log.Printf("Could not update CSR status: %v", err)
+		r.status.SetDegraded(statusmanager.CertificateSigner, "UpdateFailure",
```

Comment:
For this error, we want to retry the reconcile, so **don't** set degraded - instead, just return `err`.

Yeah, I know this is tricky.

Reply:
Good point.

Reply:
As a rule, a temporary apiserver blip doesn't need to be degraded (but it could be...). We mark degraded when there's something we'd like to report to the administrator.

Reply:
Clear as mud, I realize.